feat(pgq): Add first version of queue validator

syordanov94 opened the pull request:
Hello all

I met @kedlas at Golab 2023 and, after some talking, I proposed that maybe having some queue validations would be useful to assure that the user does not use "invalid" queue schemas.

In this PR you will find a first version of this **queue validator** that includes the following:

- A new optional attribute that enables consumer queue validation
- The _validator.go_ that is in charge of the queue validation. For now this validation only check if the queue whose name was provided, contains all the mandatory columns the **pgq** project needs to correctly function.
- Some tests agains correct and incorrect schemas in the _validator_test.go_

You will see that I have left some _TODOs_ for future additions like index validations, more precise error msgs when validating, etc.. feel free to implement any of these (maybe if I have more free time I might do some of them as well 😄)

I would appreciate any comment and suggestion you might have. I hope you find my proposal interesting ❤️ 

Thanks and great job on this project, it's very cool!

prochac commented:
Hi @syordanov94, in a first place, thank you for your contribution.

Your idea is compatible with us for sure, because we already do some limited table check here:
https://github.com/dataddo/pgq/blob/25e98a5ad8cfb9c3f00326c570ab6abf06442865/consumer.go#L292-L345

If you could modify the function and add the type check, please.

And about the `x/schema` package, it's a wish of @kedlas, but I'm a bit skeptical about it. The reason why is that we plan to introduce Terraform provider that will handle the queues' management for us. Until that, though, we use SQL migration tool. I think, no-one should use the `x/schema` package in production.
The `x` package is eXtra eXperimantal and eXpendable :smile: and also compromise between me and my CTO :smile:

syordanov94 commented:
Hey @prochac 

Oh, I didn't see this function 😅 

It basically does the same as the `Validate` function I wrote in the `validator.go` so I can just add some more logic and, for example, validate the indexes and so on. I can also test this **verifyTable** (I see there are no specific tests for it).  

Do you want validation to be mandatory? This is because I added an optional parameter in the config to enable validation. But given that it already validates it in prod, I guess it doesn't make sense to add optional validation now 😅 

Also regarding the `x/schema` you mention, I saw it in the **Readme** and that's why I used it to be honest haha. Given that it's for internal usage maybe we can use it for testing like creating "temporal" queues and destroying them once the test is completed (like I did in the validator_test.go). What do you think? 

Thanks for answering. Once I get some time I will try and "re-write" my proposal 👍

syordanov94 commented:
Hey @prochac @kedlas 

First of all, merry christmas! I hope you guys are enjoying your holidays.

I have made some updates to your `verifyTable` function, combining it with the functions I implemented in the `validator` package. 

Feel free to check it out and let me know if you find it useful.

Best Regards
Simeon

prochac commented:
@syordanov94 I do have some pending review, but it's hard to find some concentration in these pesky times :smile:

prochac reviewed (changes requested):
Finally, I reserved some time. Sorry it took so long. Holidays + tasks stacked during holidays :grimacing: 

Here are some cheap change requests. Now I will dive deeper into the other things.

Please, rebase your code. I tried and there is just a small conflict at `consumerConfig struct` in `comsumer.go`

prochac commented on validator/validator.go:
```diff
@@ -3,59 +3,195 @@ package validator
 	}
-	return ret, nil
+	return true
 }
```
The solution using for-loops without `map` is faster. Yet, both are in some sense incorrect
See here:
https://go.dev/play/p/Gy9jLUwu72G

But the incorrectness is acceptable, as the scenario is irrelevant.

prochac commented on validator/validator.go:
```diff
@@ -0,0 +1,197 @@
+		if err := rows.Scan(&indexName, &indexColumns); err != nil {
+			return nil, errors.Wrap(err, "reading index schema row of queue table")
+		}
+		indexes[indexName] = strings.Split(indexColumns, ",")
```
Here is just `","`, no space
for `"foo, bar".split(",")` => `["foo", " bar"]` and `"bar" != " bar"`

syordanov94 replied:
You are 100% correct... This was something I researched and the usage for it was to display the columns of the indexes rather than split them later on... I have made the changes, nice one 😄

prochac commented on validator/validator.go:
```diff
@@ -0,0 +1,197 @@
+const indexSelect = `
+SELECT
+    i.relname as index_name,
+    array_to_string(array_agg(a.attname), ', ') as column_names
```
Notice, you use `', '`, with space.

prochac reviewed (changes requested):
The use of `array_to_string` has some limits:
```
DROP TABLE IF EXISTS "stupid_table";
CREATE TABLE "stupid_table"
(
    "foo,bar" INT NOT NULL,
    "baz"     INT NOT NULL

);
CREATE INDEX "stupid_table_foo,bar_baz" ON "stupid_table" ("foo,bar", "baz");

SELECT
    i.relname as index_name,
    array_to_string(array_agg(a.attname), ', ') as column_names
FROM
    pg_class t,
    pg_class i,
    pg_index ix,
    pg_attribute a
WHERE
    t.oid = ix.indrelid
  and i.oid = ix.indexrelid
  and a.attrelid = t.oid
  and a.attnum = ANY(ix.indkey)
  and t.relkind = 'r'
  and t.relname = 'stupid_table'
GROUP BY
    t.relname,
    i.relname
ORDER BY
    t.relname,
    i.relname;
```

syordanov94 commented:
Hey @prochac!

Thanks for the detailed review! I have made all the mentioned changes you suggested. Please feel free to re-review! 

I ran the tests locally and everything seems ok

prochac commented:
The tests are not passing.

Is necessary to have validator in separate package and having it exported?
Imo it's ok to have it unexported in pgq root pkg. It also solves the duplicate `func openDB(t *testing.T) *sql.DB` test helper

syordanov94 commented:
Hey @prochac 

Sorry, I was missing an import in the `validator_test.go` file. Just ran them and they are passing.

Regarding the package where the validator should be located, yes it can reside in the pgq package. I added it to the validator package just because I like to split these types of features per package (maybe something more "me" 😆) because, if you have multiple validation methods (maybe some manual validation or use some external library or something) it's easier to integrate it. But pgq package works fine, I have added it!

Can you re-review? And re run the tests?

btw, the duplication of the  `func openDB(t *testing.T) *sql.DB` won't be solved by this since this functionality is not in the `pgq` package but in the `integtest` package

prochac commented:
@syordanov94 
Thank you for your contribution and your patience.
It was a very useful experience for me. The public reviews do have a bit different dynamic than internal.

kedlas commented:
Thank you @syordanov94 for contribution. If you use `pgq`, feel free to come with any new ideas. They're very welcome.
